section_inverted_index.go

```diff
@@ -0,0 +1,942 @@
+						if err != nil {
+							return nil, err
+						}
+
+						err = locEncoder.Add(docNum, loc.arrayposs...)
+						if err != nil {
+							return nil, err
+						}
+					}
+					locOffset += freqNorm.numLocs
+				}
+
+				freqNormOffset++
+
+				docTermMap[docNum] = append(
+					append(docTermMap[docNum], term...),
+					termSeparator)
+			}
+
+			tfEncoder.Close()
+			locEncoder.Close()
+
+			postingsOffset, err :=
+				writePostings(postingsBS, tfEncoder, locEncoder, nil, w, buf)
+			if err != nil {
+				return nil, err
+			}
+
+			if postingsOffset > uint64(0) {
+				err = io.builder.Insert([]byte(term), postingsOffset)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			tfEncoder.Reset()
+			locEncoder.Reset()
+		}
+
+		err = io.builder.Close()
+		if err != nil {
+			return nil, err
+		}
+
+		// record where this dictionary starts
+		dictOffsets[fieldID] = uint64(w.Count())
+
+		vellumData := io.builderBuf.Bytes()
+
+		// write out the length of the vellum data
+		n := binary.PutUvarint(buf, uint64(len(vellumData)))
+		_, err = w.Write(buf[:n])
+		if err != nil {
+			return nil, err
+		}
+
+		// write this vellum to disk
+		_, err = w.Write(vellumData)
+		if err != nil {
+			return nil, err
+		}
+
+		// reset vellum for reuse
+		io.builderBuf.Reset()
+
+		err = io.builder.Reset(&io.builderBuf)
+		if err != nil {
+			return nil, err
+		}
+
+		// write the field doc values
+		// NOTE: doc values continue to use legacy chunk mode
+		chunkSize, err := getChunkSize(LegacyChunkMode, 0, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		fdvEncoder := newChunkedContentCoder(chunkSize, uint64(len(io.results)-1), w, false)
+		if io.IncludeDocValues[fieldID] {
+			for docNum, docTerms := range docTermMap {
+				if len(docTerms) > 0 {
+					err = fdvEncoder.Add(uint64(docNum), docTerms)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+			err = fdvEncoder.Close()
+			if err != nil {
+				return nil, err
+			}
+
+			fdvOffsetsStart[fieldID] = uint64(w.Count())
+
+			_, err = fdvEncoder.Write()
+			if err != nil {
+				return nil, err
+			}
+
+			fdvOffsetsEnd[fieldID] = uint64(w.Count())
+			fdvEncoder.Reset()
+		} else {
+			fdvOffsetsStart[fieldID] = fieldNotUninverted
+			fdvOffsetsEnd[fieldID] = fieldNotUninverted
+		}
+
+		fieldStart := w.Count()
+
+		// todo: uvarint these offsets
+		err = binary.Write(w, binary.BigEndian, fdvOffsetsStart[fieldID])
+		if err != nil {
+			return nil, err
+		}
+
+		err = binary.Write(w, binary.BigEndian, fdvOffsetsEnd[fieldID])
+		if err != nil {
+			return nil, err
+		}
+
+		err = binary.Write(w, binary.BigEndian, dictOffsets[fieldID])
+		if err != nil {
+			return nil, err
+		}
+
+		io.fieldAddrs[fieldID] = fieldStart
+	}
+
+	return dictOffsets, nil
+}
+
+func (io *invertedIndexOpaque) process(field index.Field, fieldID uint16, docNum uint64) {
+	if !io.init && io.results != nil {
+		io.prepareDicts()
+		io.init = true
+	}
+
+	if fieldID == ^uint16(0) {
```

Comment:
Would you add some commentary around when fieldID is set to this - which should be math.MaxUint16.

Reply:
yeah, i'll be adding commentary to certain portions of the changes in the coming commits.